rust/src/initramfs.rs

```diff
@@ -0,0 +1,179 @@
+    if let Some(c) = cancellable {
+        let _ = c.set_error_if_cancelled()?;
+    }
+    let meta = d.metadata(path).context("stat")?;
+    match meta.simple_type() {
+        SimpleType::Symlink | SimpleType::File => {}
+        SimpleType::Dir => {
+            for e in d.list_dir(path).context("readdir")? {
+                let e = e.context("readdir")?;
+                let name = e.file_name();
+                let name = match name.to_str() {
+                    Some(n) => n,
+                    None => anyhow::bail!("Invalid UTF-8 name {}", name.to_string_lossy()),
+                };
+                let subpath = format!("{}/{}", path, name);
+                list_files_recurse(&d, &subpath, filelist, cancellable)?;
+            }
+        }
+        _ => anyhow::bail!("Invalid non-regfile/symlink/directory: {}", path),
+    }
+    filelist.insert(path.to_string());
+    Ok(())
+}
+
+fn gather_filelist<P: glib::IsA<gio::Cancellable>>(
+    d: &openat::Dir,
+    input: &HashSet<String>,
+    cancellable: Option<&P>,
+) -> Result<BTreeSet<String>> {
+    let mut filelist: BTreeSet<String> = BTreeSet::new();
+    for file in input {
+        let file = match file.strip_prefix("/etc/") {
+            Some(f) => f,
+            None => anyhow::bail!("Invalid non-/etc path: {}", file),
+        };
+        list_files_recurse(&d, file, &mut filelist, cancellable)
+            .with_context(|| format!("Adding {}", file))?;
+    }
+    Ok(filelist)
+}
+
+fn generate_initramfs_overlay<P: glib::IsA<gio::Cancellable>>(
+    root: &openat::Dir,
+    files: &HashSet<String>,
+    cancellable: Option<&P>,
+) -> Result<fs::File> {
+    let filelist = {
+        let etcd = root.sub_dir("etc")?;
+        gather_filelist(&etcd, files, cancellable)?
+    };
+    let out_tmpf = std::rc::Rc::new(tempfile::tempfile()?);
+    {
+        let cmd = Exec::cmd("cpio")
+            .args(&[
+                "--create",
+                "--format",
+                "newc",
+                "--quiet",
+                "--reproducible",
+                "--null",
+            ])
+            .cwd(format!("/proc/self/fd/{}", root.as_raw_fd()))
+            | Exec::cmd("gzip").arg("-1");
+        let mut children = cmd
+            .stdin(subprocess::Redirection::Pipe)
+            .stdout(subprocess::Redirection::RcFile(out_tmpf.clone()))
+            .popen()
+            .context("spawn")?;
+        {
+            let mut cstdin = std::io::BufWriter::new(children[0].stdin.take().expect("stdin"));
+            for f in filelist {
+                cstdin.write_all(b"etc/")?;
+                cstdin.write_all(f.as_bytes())?;
+                let nul = [0u8];
+                cstdin.write_all(&nul)?;
+            }
+            cstdin.flush()?;
+        }
+        for mut child in children {
+            let status = child.wait()?;
+            if !status.success() {
+                anyhow::bail!("Failed to generate cpio archive");
```

Comment:
Can we print the command of the child that failed here?

Reply:
Not easily, doesn't look like there's a way to go from a `Child` back to `Exec` to run [to_cmdline_lossy](https://docs.rs/subprocess/0.2.6/subprocess/struct.Exec.html#method.to_cmdline_lossy) - I would agree this would be a good thing in general but in this specific instance if anything fails it's going to be cpio, and it should be obvious.